Add address rewards metadata

Doy-lee opened the pull request:
This adds some new fields to the rewards DB.

*Batched Accrued Rewards*
- lifetime_locked_stakes
- lifetime_unlocked_stakes
- lifetime_liquidated_stakes
- lifetime_rewards

These numbers expose some more metadata on what exactly is the composition of the `amount` value (which is rewards + unlocked stake) into separate fields. You can calculate how many tokens are actively locked in the network `(lifetime locked - (lifetime unlocked + lifetime liquidated))` or see how many rewards an address has earnt (`lifetime_rewards`) or how much has been liquidated from the address (`lifetime_liquidated`).

I've updated the `get_accrued_rewards` to return a struct now which has all this metadata plus more. It's basically all documented in the API

```cpp
struct wallet_info {
        uint64_t height;  // Height at which the wallet info was retrieved for

        // True if the wallet has participated in the network or not. When false all values are
        // zeroed (because the wallet has not earnt or spent any tokens on the network).
        bool found;

        // For OXEN addresses, rewards that has been accrued but not paid out to the address yet.
        //
        // For ETH addresses, lifetime rewards _and_ unlocked stakes for the address. (Note that,
        // unlike Oxen addresses, these rewards never reset to zero; but rather the rewards contract
        // keeps track of the current paid and current total and pays out the difference).
        //
        // For ETH addresses, this can be derived by
        // `lifetime_unlocked_stakes + lifetime_rewards - lifetime_liquidated_stakes`
        cryptonote::reward_money amount;

        // For ETH addresses _only_, the total amount of tokens that were locked by this wallet and
        // so forth for lifetime unlocked, rewards and liquidated amounts.
        cryptonote::reward_money lifetime_locked_stakes;
        cryptonote::reward_money lifetime_unlocked_stakes;
        cryptonote::reward_money lifetime_rewards;
        cryptonote::reward_money lifetime_liquidated_stakes;

        // For ETH addresses _only_, the amount of tokens currently locked in the network (e.g.
        // staked in a node). This is defined as `lifetime locked - lifetimed unlocked` and
        // precalculated for convenience. This number includes `timelocked_stakes`.
        cryptonote::reward_money locked_stakes;

        // For ETH addresses _only_, the amount of tokens awaiting to be unlocked from the network
        // (e.g. an exit has been processed and the stake is under a time lock before being
        // claimable by the address)
        cryptonote::reward_money timelocked_stakes;
    };
```

The way this is achieved is by simply storing the result of the ETH transactions that we confirm and witness and feed it down into the rewards DB where it enumerates them together. These new fields don't kick in until the ETH hardfork so they're all gated and should return 0 if queried before then. This way collecting this metadata only needs to happen on the ETH code path instead of also requiring to work on the OXEN path.

Since it's gated, this will only trigger a rescan on stagenet which has 50k blocks. Since the new fields are written on the ETH hardfork there'll be no rescan needed for mainnet. It'll just gracefully start enumerating the fields when the hardfork elapses.

This does overlap with the Oxen 10 backwards compat fix that was recently patched. I had to patch ontop again to preserve the old behaviour. For Oxen 11, I just made it support parsing the new `wallet_info` struct that we return in the RPC call.

The `get_accrued_rewards` now returns something like this 

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "balances": [
      {
        "address": "0x26b0227617159797b9301a8EA6FB264f17d37Cb4",
        "amount": 181398273044375,
        "found": true,
        "lifetime_liquidated_stakes": 240000000012,
        "lifetime_locked_stakes": 270774959746030,
        "lifetime_rewards": 8985313298357,
        "lifetime_unlocked_stakes": 172652959746030,
        "locked_stakes": 98122000000000,
        "rewards": 8985313298357,
        "timelocked_stakes": 18122000000000
      }
    ],
    "height": 50062,
    "status": "OK"
  }
}
```

Doy-lee committed:
Early continue in unconfirmed TXs loop to allow unindent of code block

Doy-lee committed:
Add non-consensus locked stakes field to rewards DB

Doy-lee committed:
Grab timelocked SESH stakes from delayed payments

Doy-lee committed:
Update docs for get_accrued_rewards

Doy-lee committed:
Update get_accrued_rewards to emit wallet_info data

Doy-lee committed:
Correctly calculate timelocked stakes and update the RPC docs

Doy-lee committed:
Remove unused unlocked stakes

Doy-lee committed:
Fix Oxen 10 backwards compat get batched rewards to handle new wallet_info output

Doy-lee committed:
Add address to docs for accrued rewards, gate new debug checks behind HF21

Doy-lee committed:
Gate lifetime rewards metadata behind HF21

Doy-lee committed:
Formatting for adding-rewards-metadata PR

Doy-lee committed:
Fix CalculateRewards, to_coin in EXPECT divides by 0, so base amount needs to be in DB units

darcys22 commented:
Nothing in consensus or in the chain should be affected by this right? Like this should sync independently and then we have a machine that will output the right locked amounts?

So we could test this by simply attaching it to the portal backend server and using these new RPC endpoints?

Doy-lee committed:
Rebase onto dev w/ reward money changes

Doy-lee committed:
Handle locked stakes and rewards generated on the migration

The rewards metadata PR was written before the HF21 transition code was
merged into the mainline branch. In the HF21 transition code, we have
some special one-off code that bootstraps the new network and converts
$OXEN->$SESH stakes whilst also generating a bunch of new bonus $SESH
tokens as part of the swap bonus program.

These new events do not go through the normal route of being emitted as
a smart contract event and witnessed on-chain which is what the rewards
metadata PR was written to expect. This means that the new stakes and
bonus rewards aren't accounted for and would lead to out-of-sync values
recorded in the SQL DB.

Verification of the SQL DB is only enabled in debug mode so nodes will
happily accept these values. The staking portal which uses these values
for visualisation purposes would go of sync. This PR fixes that by
reusing some functions that are expected to be called when the SQL DB
encounters a new locked stake or when a contributor receives a reward
thus synchronising the bootstrapped $SESH into the DB.

Doy-lee committed:
Make mocknet work w/ rewards metadata changes

tewinget commented on src/cryptonote_core/sent_transition/sent_transition.h:
```diff
@@ -74,6 +71,8 @@ void transition(
         cryptonote::BlockchainSQLite& sql,
-        network_type net);
+        network_type net,
+        service_nodes::block_add_result& add_result,
```
Unless I missed something, the two places which call this function create a container of this type and pass it in, but then don't do anything with the container.

Doy-lee replied:
It is used, we populate it with the new locked stakes and rewards. For the HF21 dry-run this isn't used, but in the main code path it's used in a functional-esque manner where updating the SNL produces these lists of stakes.

The lists of stakes are fed into the SQL layer which then ingests it and updates the tables. Essentially boils down to something like

```cpp
blockchain.update();
block_add_result add_result = snl.update();
sql_db.update(add_result);
```

jagerman commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -526,24 +595,167 @@ std::pair<int, std::string> BlockchainSQLite::get_address_str(
+        eth::address eth_addr = {};
+        bool is_eth = tools::try_load_from_hex_guts(address, eth_addr);
+
+        int64_t amount = std::get<0>(*metadata);
+        int64_t lifetime_locked_stakes = std::get<1>(*metadata);
+        int64_t lifetime_unlocked_stakes = std::get<2>(*metadata);
+        int64_t lifetime_liquidated_stakes = std::get<3>(*metadata);
+        int64_t lifetime_rewards = std::get<4>(*metadata);
```
Modernize:
```suggestion
        auto [amount, lifetime_locked_stakes, lifetime_unlocked_stakes, lifetime_liquidated_stakes, lifetime_rewards] = *metadata;
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/sent_transition/sent_transition.cpp:
```diff
@@ -811,9 +821,42 @@ void transition(
+        // Record the stake metadata to the `block_add_result`. The SQL DB will process these when
+        // it's update set is triggered
+        for (size_t contrib_index = 0; contrib_index < it.sn_info->contributors.size(); contrib_index++) {
+            auto contrib_it = it.sn_info->contributors[contrib_index];
```
Reference?

Also using "it" as a name (or part of a name) for a concrete value rather than an iterator feels a bit off.

Doy-lee replied:
Referenced and renamed to `contrib`

jagerman commented on src/cryptonote_core/sent_transition/sent_transition.cpp:
```diff
@@ -811,9 +821,42 @@ void transition(
+            assert(!it.zombie && "Contributors should only be populated on non-zombie nodes");
+            assert(contrib_it.ethereum_address);
+            assert(contrib_it.address == cryptonote::account_public_address{});
+            service_nodes::eth_stake stake = {
+                    .sn = crypto::ed25519_public_key{it.pkey},
+                    .addr = contrib_it.ethereum_address,
+                    .amount = cryptonote::reward_money::coin_amount(contrib_it.amount),
+                    .liquidation = cryptonote::reward_money{},
+                    .block_height = static_cast<uint32_t>(snl_state.height),
+                    .tx_index = synthetic_tx_index++,
+                    .contributor_index = static_cast<uint32_t>(contrib_index),
+            };
+            add_result.locked_stakes.push_back(stake);
```
An emplace_back here would seem cleaner than a make-and-copy:
```suggestion
            add_result.locked_stakes.emplace_back(
                    crypto::ed25519_public_key{it.pkey},
                    contrib_it.ethereum_address,
                    cryptonote::reward_money::coin_amount(contrib_it.amount),
                    cryptonote::reward_money{},
                    static_cast<uint32_t>(snl_state.height),
                    synthetic_tx_index++,
                    static_cast<uint32_t>(contrib_index));
```
or alternatively:
```suggestion
            auto& s = add_result.locked_stakes.emplace_back();
            s.sn = crypto::ed25519_public_key{it.pkey};
            s.addr = contrib_it.ethereum_address;
            s.amount = cryptonote::reward_money::coin_amount(contrib_it.amount);
            s.liquidation = cryptonote::reward_money{};
            s.block_height = static_cast<uint32_t>(snl_state.height);
            s.tx_index = synthetic_tx_index++;
            s.contributor_index = static_cast<uint32_t>(contrib_index);
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -2004,16 +2105,19 @@ bool service_node_list::state_t::process_confirmed_event(
-        return false;
+        return result;
     }
 }
```
A tiny cleanup here would be to delete the two `return result;`s above (in the try, and in catch) and add a single `return result;` at the bottom of the function.

Doy-lee replied:
Done

jagerman commented on src/cryptonote_basic/cryptonote_format_utils.h:
```diff
@@ -300,14 +300,14 @@ constexpr std::string_view get_unit() {
-        size_t decimal_places = oxen::DISPLAY_DECIMAL_POINT,
-        bool strip_zeros = false);
+        bool strip_zeros = false,
+        size_t decimal_point = oxen::DISPLAY_DECIMAL_POINT);
```
There are some uses of this in the codebase that are now passing reverse arguments here (I think they may be recent dev additions).  Adding these here would make those errors:
```C++
// Guard against passing arguments in wrong order:
std::string print_money(uint64_t, size_t, bool) = delete;
std::string print_money(uint64_t, size_t) = delete;
```

Doy-lee replied:
That's a great trick, works well- used it to find the bad call-sites but then after that I've ultimately made `strip_zeros` a type safe enum and be explicit and avoid the delete lines being a permanent fixture.

jagerman commented on src/cryptonote_core/sent_transition/sent_transition.cpp:
```diff
@@ -464,14 +465,13 @@ void transition(
                 eth_addr,
-                val);
-        converted_rewards[oxen_addr] = val;
+                cryptonote::print_money(val.amount.to_coin()));
```
Is there a reason here (and several other places using the same pattern) to not use just `val.amount` as the format argument, to format it via reward_money::to_string()?  `cryptonote::print_money(val.amount.to_coin())` as an argument means it *always* has to do string formatting, even if the log statement isn't going to be displayed, while passing a formattable type only formats it when needed.  The other difference is that print_money will drop off the subatomic digits, but I don't really see that as being needed in most of these log statements.

Doy-lee replied:
No, I didn't realise we had a `to_string()` implementation for it, so this can be dropped here.

jagerman commented on src/cryptonote_core/sent_transition/sent_transition.cpp:
```diff
@@ -784,8 +784,18 @@ void transition(
+    {
+        sql.db.exec("DELETE FROM batched_payments_accrued");
+        cryptonote::block_payments rewards_payments;
+        for (auto it : unallocated) {
```
```suggestion
        for (const auto& [eth_addr, amt] : unallocated) {
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
+        std::unordered_map<eth::address, stake_history> eth_to_locked_stakes;
+
+        // NOTE: Enumerate SNL
+        for (auto it : state.service_nodes_infos) {
+            for (auto contrib_it : it.second->contributors) {
```
These (and several similar loops below) are copying each element as they iterate.
```suggestion
        for (const auto& [pk, sni] : state.service_nodes_infos) {
            for (const auto& contributor : sni->contributors) {
```

jagerman replied:
I think the misunderstanding here (and in the other cases) is that `auto` has to be opt-in on capturing a reference and so these loop iterations are copying:

```c++
#include <iostream>
#include <vector>

struct A {
    A() { std::cout << "A ctor\n"; }
    A(A&&) { std::cout << "A move ctor\n"; }
    A(const A&) { std::cout << "A copy ctor\n"; }
    A& operator=(const A&) { std::cout << "A copy assign\n"; return *this; }
    A& operator=(A&&) { std::cout << "A move assign\n"; return *this; }
};

int main() {
    std::vector<A> v;
    v.resize(10);
    for (auto x : v) {}
}
```
```
$ ./a.out
A ctor
A ctor
A ctor
A copy ctor
A copy ctor
A copy ctor
```
but with `for (auto& x : v) {}` does not:
```
A ctor
A ctor
A ctor
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
+
+        // NOTE: Enumerate nodes in the recently removed list (considered locked until exited)
+        for (auto it : state.recently_removed_nodes) {
+            for (auto contrib_it : it.info.contributors) {
```
As above.

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
+        request.type = cryptonote::BlockchainSQLite::delayed_payments_type::all;
+
+        cryptonote::block_payments locked_payments = db.get_delayed_payments(request);
+        for (auto it : locked_payments) {
```
```suggestion
        for (const auto& [addr, payment] : locked_payments) {
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
+        }
+
+        // NOTE: Walk the locked stakes per ETH address and verify the numbers in the rewards DB
+        for (auto it : eth_to_locked_stakes) {
```
```suggestion
        for (const auto& [addr, history] : eth_to_locked_stakes) {
```

Doy-lee replied:
Done

jagerman commented on src/cryptonote_core/service_node_list.cpp:
```diff
@@ -291,6 +291,104 @@ static uint64_t min_recent_height(cryptonote::network_type nettype, uint64_t hei
+                        cryptonote::print_money(it.second.total),
+                        cryptonote::print_money(wallet_info.locked_stakes.to_coin()));
+
+                for (auto stake_it : it.second.stakes) {
```
```suggestion
                for (const auto& stake : hist) {
```

Doy-lee replied:
Done

jagerman commented on src/wallet/wallet2.cpp:
```diff
@@ -15274,9 +15274,19 @@ void wallet2::refresh_batching_cache() {
+    for (auto it : res["balances"]) {
+        nlohmann::json::iterator address = it.find("address");
+        nlohmann::json::iterator amount = it.find("amount");
+        assert(address != it.end());
+        assert(amount != it.end());
```
These need to be checked in non-debug builds, too.

Doy-lee replied:
Done

jagerman commented on src/rpc/core_rpc_server.cpp:
```diff
@@ -3881,47 +3881,82 @@ void core_rpc_server::invoke(ONS_RESOLVE& resolve, rpc_context) {
+        for (auto& it : array) {
+            nlohmann::json::iterator address = it.find("address");
+            nlohmann::json::iterator amount = it.find("amount");
+            assert(address != it.end());
+            assert(amount != it.end());
```
These need to be checked in non-debug builds too.

Doy-lee replied:
Done

jagerman commented on src/wallet/wallet2.cpp:
```diff
@@ -15274,9 +15274,19 @@ void wallet2::refresh_batching_cache() {
+        assert(address->is_string());
+        assert(amount->is_number_unsigned());
+
+        auto address_str = address->get_ref<const std::string&>();
```
This is getting a reference but then making a copy, which makes the `get_ref` a bit weird.
```suggestion
        auto& address_str = address->get_ref<const std::string&>();
```

Doy-lee replied:
Done

jagerman commented on src/wallet/wallet2.cpp:
```diff
@@ -15274,9 +15274,19 @@ void wallet2::refresh_batching_cache() {
-    auto records = res["balances"].get<std::unordered_map<std::string, uint64_t>>();
     batching_records_cache.clear();
-    batching_records_cache = std::move(records);
+    for (auto it : res["balances"]) {
```
```suggestion
    for (const auto& it : res["balances"]) {
```

Doy-lee replied:
Done

Doy-lee committed:
Fix decimal_points/strip_zeros parameter swap being misused

Use a type-safe enum and make the call sites explicit state what
behaviour they want to avoid decimal points being passed as strip zeros
and being coerced to bools.

Doy-lee committed:
Code review: for-each fixes, non-debug checks, structured bindings

Doy-lee committed:
Linting pass in mocknet
